zlib-rs/src/crc32.rs

```diff
@@ -0,0 +1,133 @@
+
+const fn build_crc32_table<const T: usize, const W: usize>(stride: usize) -> [[u32; T]; W] {
+    let mut arr = [[0u32; T]; W];
+    let mut i = 0;
+    while i < W {
+        let mut j = 0;
+        while j < T {
+            let mut c = j;
+            let mut k = 0;
+            while k < 8 * (W * stride - i) {
+                if c & 1 != 0 {
+                    c = CRC32_LSB_POLY ^ (c >> 1);
+                } else {
+                    c >>= 1;
+                }
+                k += 1;
+            }
+            arr[i][j] = c as u32;
+            j += 1;
+        }
+        i += 1;
+    }
+    arr
+}
+
+fn crc32_naive_inner(data: &[u8], start: u32) -> u32 {
+    data.iter().fold(start, |crc, val| {
+        let crc32_lsb = crc.to_le_bytes()[0];
+        CRC32_BYTE_TABLE[0][usize::from(crc32_lsb ^ *val)] ^ (crc >> 8)
+    })
+}
+
+pub fn crc32_naive(data: &[u8]) -> u32 {
+    crc32_naive_inner(data, START_CHECKSUM) ^ FINAL_XOR
+}
+
+fn crc32_words_inner(words: &[usize], start: u32, per_word_crcs: &[u32]) -> u32 {
+    words.iter().enumerate().fold(start, |crc, (i, word)| {
+        let mut word_crc = 0u32;
+        let value = *word ^ (crc ^ per_word_crcs.get(i).unwrap_or(&0)) as usize;
+        for (i, b) in value.to_le_bytes().into_iter().enumerate() {
+            word_crc ^= CRC32_WORD_TABLE[i][usize::from(b)];
+        }
+
+        // Per Listing 4 in Kadatch and Jenkins, this step
+        // is needed in case size_of(usize) < size_of(u32)
+        if std::mem::size_of_val(&word_crc) <= W {
+            word_crc
+        } else {
+            (crc >> 8) ^ word_crc
+        }
+    })
+}
+
+pub fn crc32_words(data: &[u8]) -> u32 {
+    // Get a word-aligned sub-slice of the input data
+    let (prefix, words, suffix) = unsafe { data.align_to::<usize>() };
+
+    let crc = crc32_naive_inner(prefix, START_CHECKSUM);
+    let crc = crc32_words_inner(words, crc, &[]);
+    crc32_naive_inner(suffix, crc) ^ FINAL_XOR
+}
+
+pub fn crc32_interleaved(data: &[u8]) -> u32 {
+    // Get a word-aligned sub-slice of the input data
+    let (prefix, words, suffix) = unsafe { data.align_to::<usize>() };
+
+    let mut crcs = [0u32; N + 1];
+    crcs[0] = crc32_naive_inner(prefix, START_CHECKSUM);
+
+    // TODO: this would normally use words.chunks_exact(N), but
+    // we need to pass the last full block to crc32_words_inner
+    let blocks = words.len() / N;
+    let stride_blocks = blocks.saturating_sub(1);
+    for i in 0..stride_blocks {
+        // Load the next N words.
+        let mut buffer: [usize; N] =
+            std::array::from_fn(|j| usize::to_le(words[i * N + j] ^ (crcs[j] as usize)));
+
+        for j in 0..N {
+            // TODO: this doesn't compile
+            // Move overflow bits into the "next" word.
+            //if std::mem::size_of_val(&crcs[i]) > W {
+            //    crcs[i + 1] ^= crcs[i] >> (8 * W);
+            //}
+            crcs[j] = 0;
+        }
```

Comment:
I'm not sure what is or should be happening here. Assuming the current logic is correct, this can be written at `crcs[0..N].fill(0)`.

Reply:
I'm following the algorithm from https://github.com/zlib-ng/zlib-ng/blob/develop/doc/crc-doc.1.0.pdf, and they handle two cases here:
* The size of the CRC (`D` in the paper) is larger than the word size
* The CRC is smaller than the word size.

Since we use `u32` for the CRC and `usize` for the word, we could ignore the first case and assert that `u32 <= usize`.